microprofile/cloud/common/src/main/java/io/helidon/microprofile/cloud/common/CommonCloudFunction.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.microprofile.cloud.common;
+
+import java.util.Optional;
+import java.util.logging.Logger;
+
+import javax.enterprise.inject.spi.CDI;
+
+import io.helidon.microprofile.cdi.Main;
+
+/**
+ * Contains common functionality to be extended by every cloud implementations.
+ *
+ * @param <T> the delegate type
+ */
+public abstract class CommonCloudFunction<T> {
+
+    /**
+     * First time is invoked it initializes Helidon and creates a delegated instance.
+     * Next invocations will reuse the same instance.
+     *
+     * @return the implementation of specific cloud interface
+     */
+    @SuppressWarnings("unchecked")
+    protected T delegate() {
+        return (T) LazyHelidonInitializer.DELEGATE;
+    }
+
+    /**
+     * Avoids Helidon is initialized in case CommonCloudFunction is never used.
+     *
+     */
+    private static class LazyHelidonInitializer {
```

Comment:
Mabe [io.helidon.common.LazyValue](https://github.com/oracle/helidon/blob/785ce38ca06b268d16e387fb6498aaaa890695cc/common/common/src/main/java/io/helidon/common/LazyValue.java) could help here

Reply:
I think current one performs a little bit better.

LazyValueImpl has some MethodHandles.lookup().findVarHandle and some lines of code related to get the value in a thread safety manner.

This 'MethodHandles.lookup().findVarHandle' is going to be executed every time there is a cold start. In lambdas it will happen quite often.

I think LazyValueImpl is better for some java processes that are running for longer time and for general usage.